### ISSUE TITLE
AutoTrackRaymarine PIM upgrade ...failing

### ISSUE THREAD

**Issue by rgleason** [OpenCPN/plugins#511]
Trying to update to work with PIM. This has been pushed to  AutoTrackRaymarine 0.3.1.1 uses api 117 [branch master](https://github.com/rgleason/AutoTrackRaymarine_pi) by Douwe Fokkema failing due

`"..unresolved external symbol "public: virtual __thiscall PlugInChartBaseGLPlus2"`

I am stuck on these failures shown in the output.txt
[output.txt](https://github.com/OpenCPN/plugins/files/7611678/output.txt)

These unresolved external sybols  appear to be associated with the  api 117 extension
...AutoTrackRaymarine_pi\libs\ocpn-api\ocpn_plugin.h

```
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\libs\ocpn-api\ocpn_plugin.h (10 hits)
	Line 878: // PlugInChartBaseExtended
	Line 882: class DECL_EXP PlugInChartBaseExtended : public PlugInChartBase {
	Line 884:   PlugInChartBaseExtended();
	Line 885:   virtual ~PlugInChartBaseExtended();
	Line 921: // PlugInChartBaseExtendedPlus2
	Line 922: //  Derived from PlugInChartBaseExtended, add additional extended chart support
	Line 926: class DECL_EXP PlugInChartBaseExtendedPlus2 : public PlugInChartBaseExtended {
	Line 926: class DECL_EXP PlugInChartBaseExtendedPlus2 : public PlugInChartBaseExtended {
	Line 928:   PlugInChartBaseExtendedPlus2();
	Line 929:   virtual ~PlugInChartBaseExtendedPlus2();
```

How can I get this resolved?

**Comment by rgleason** [comment]
What I have found is that I cannot use the current version of ocpn_plugin.h,   which has additions to the 117 section of the api, I have to use the original ocpn_plugin.h  file.   I don't know why yet.  Perhaps I should compare them.

When I use the original ocpn_plugin.h the plugin does compile.

**Comment by rgleason** [comment]
When I compare the two ocpn_plugin.h files, the upper part seems to be identical except for spaces. Then there has been reorganization in the order for the middle part so it is difficult to determine, but for the end after API 117 the initial section appears to be identical and Douwe's original ocpn_plugin.h seems to be just missing the part after  //Extended Waypoint manipulation API.

So why doesn't the current ocpn_plugin.h file work?

**Comment by rgleason** [comment]
I have tried comparing the two files and copying the upper part over the original ocpn_plugin.h that Douwe was using., however it does not have the Extended Waypoint manipulation section at the bottom.
This file does allow AutoTrackRaymarine to compile and load into Windows.

While doing this I found the  "PlugInChartBaseExtendedPlus2"  that was causing error notices earlier.

**Comment by rgleason** [comment]
Next I tried using the entire updated ocpn_plugin.h file from the current version of Opencpn which includes the section below the Extended Waypoint Manipulation comment.  

It fails with the same unresolved external symbols issue as initially stated  earlier, see output below
[output.txt.docx](https://github.com/OpenCPN/plugins/files/7618795/output.txt.docx)


Why? and how can I fix this?

**Comment by rgleason** [comment]
Now I copied the extended waypoint part and it fails, so I removed that and I am getting the same error again 

```
C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.exp
NGT1Read.obj : error LNK2001: unresolved external symbol "public: virtual __thiscall PlugInChartBaseExtendedPlus2::~PlugInChartBaseExtendedPlus2(void)" (??1PlugInChartBaseExtendedPlus2@@UAE@XZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
SerialPort.obj : error LNK2001: unresolved external symbol "public: virtual __thiscall PlugInChartBaseExtendedPlus2::~PlugInChartBaseExtendedPlus2(void)" (??1PlugInChartBaseExtendedPlus2@@UAE@XZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
```

So I am going to have to go back and use douwe's original ocpn_plugin.h file.

**Comment by bdbcat** [comment]
Rick...
You must be linking against an older version of OCPN that did not have the 117 definitions included.
Are you building locally?  Can you show the full buildlog?

**Comment by rgleason** [comment]
Building locally with Windows 10,  doing "standalone" builds of the [AutoTrackRaymarine plugin](https://github.com/rgleason/AutoTrackRaymarine_pi)  Master branch using  [libs/ocpn-api/ocpn-plugin.h](https://github.com/rgleason/AutoTrackRaymarine_pi/tree/master/libs/ocpn-api)

O was cloned and built about 1 month ago. Will update and then build the plugin again with that ocpn-plugin.h and report back.

**Comment by rgleason** [comment]
Using the ocpn-plugin.h from OpenCPN freshly downloaded, and copied to source/AutoTrackRaymarine/libs/ocpn-api/ocpn-plugin.h  

Then ran cmake and tried to build locally. Attached are the bash prompt output and the the output.txt file.
The errors conform to the above errors
[bash-prompt-output.txt](https://github.com/OpenCPN/plugins/files/7622169/bash-prompt-output.txt)
[output.txt](https://github.com/OpenCPN/plugins/files/7622177/output.txt)
.

**Comment by rgleason** [comment]
The version that works is OLD-opcn-plugins.h in that directory.

**Comment by bdbcat** [comment]
Try updating the opencpn.lib file (AutoTrackRaymarine_pi/libs/ocpn-api/msvc/opencpn.lib)
This will be necessary for the latest plugin API.

btw, why are you using API 117?  One should use the _LOWEST_ API number that satisfies the plugin's requirements.  I expect 116 would work as well, for this plugin.

this one supports API 117:
https://www.dropbox.com/s/60u4qv95w8slec0/opencpn.lib?dl=0

**Comment by rgleason** [comment]
This is using 117api, it is for Douwe's plugin for Raymarine because it needed greater accuracy.

**Comment by rgleason** [comment]
Thanks, I never thought of those two files as being coupled.

**Comment by rgleason** [comment]
To be sure about ocpn_plugins.h, I copied it again from opencpn/include/  pulled yesterday.
Now getting 
[output.txt.docx](https://github.com/OpenCPN/plugins/files/7626070/output.txt.docx)

Search "opencpn_plugin_117" (4 hits in 3 files of 127 searched)
```
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.cpp (1 hit)
	Line 75:   : opencpn_plugin_117(ppimgr)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.h (1 hit)
	Line 74: class AutoTrackRaymarine_pi : public wxEvtHandler, public opencpn_plugin_117
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\OLD\OLD-ocpn_plugin.h (2 hits)
	Line 570: class DECL_EXP opencpn_plugin_117 : public opencpn_plugin_116
	Line 573:     opencpn_plugin_117(void *pmgr);

Perhaps something is not configured right.
```

**Comment by rgleason** [comment]
src\AutoTrackRaymarine_pi.cpp
```
AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *ppimgr)
  : opencpn_plugin_117(ppimgr)
{
  // Create the PlugIn icons
  initialize_images();

```

src\AutoTrackRaymarine_pi.h 

```
class AutoTrackRaymarine_pi : public wxEvtHandler, public opencpn_plugin_117
{
  //friend InfoDialog;
public:
  AutoTrackRaymarine_pi(void *ppimgr);
  int Init(void);
  bool DeInit(void);

  int GetAPIVersionMajor();
  int GetAPIVersionMinor();
  int GetPlugInVersionMajor();
```

**Comment by rgleason** [comment]
Here is the problem

Search "GetPlugInVersionPatch" (3 hits in 1 file of 127 searched)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\OLD\OLD-ocpn_plugin.h (3 hits)
	Line 578:     virtual int GetPlugInVersionPatch();
	Line 583:     /** Pre-release tag version part, see GetPlugInVersionPatch() */
	Line 586:     /** Build version part  see GetPlugInVersionPatch(). */

**Comment by rgleason** [comment]
Here are some output logs

[output-cmd-first.txt.docx](https://github.com/OpenCPN/plugins/files/7626176/output-cmd-first.txt.docx)

[output.txt.docx](https://github.com/OpenCPN/plugins/files/7626171/output.txt.docx)

**Comment by rgleason** [comment]
In ocpn_plugin.h  at line 547

```
class DECL_EXP opencpn_plugin_117 : public opencpn_plugin_116 {
public:
  opencpn_plugin_117(void *pmgr);
  /*
   * Forms a semantic version together with GetPlugInVersionMajor() and
   * GetPlugInVersionMinor(), see https://semver.org/
   */
  virtual int GetPlugInVersionPatch();
```

so why   "opencpn_plugin_117::opencpn_plugin_117"?

**Comment by rgleason** [comment]
Search "opencpn_plugin_116" (6 hits in 1 file of 125 searched)
```
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\.vs\src\v15\Browse.VC.db (6 hits)
	Line 9145: 
	Line 9149: 
	Line 9151: 
	Line 482939: 	ñ	á	Ñ	Á	±	¡	‘		q	a	Q	A	1	!		ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡qaQA1!ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡‘qaQA1!°¡’ƒteVG8)üíÞÏ
	Line 482939: 	ñ	á	Ñ	Á	±	¡	‘		q	a	Q	A	1	!		ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡qaQA1!ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡‘qaQA1!°¡’ƒteVG8)üíÞÏ
	Line 759315: ~LATLONGÄ~weak_ptr®~vectorT~vectorT#~unique_ptrÀ#~unique_ptra!~type_infoQC#~shared_ptrw1~opencpn_plugin_19¡1~opencpn_plugin_18‘1~opencpn_plugin_17ƒ1~opencpn_plugin_16x3~opencpn_plugin_116Ø3~opencpn_plugin_115Ó3~opencpn_plugin_114Î3~opencpn_plugin_113Ã3~opencpn_plugin_112´3~opencpn_plugin_111¯3~opencpn_plugin_110¨+~opencpn_plugin=/~nested_exception$b
Search "opencpn_plugin_116" (6 hits in 1 file of 125 searched)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\.vs\src\v15\Browse.VC.db (6 hits)
	Line 9145: 
	Line 9149: 
	Line 9151: 
	Line 482939: 	ñ	á	Ñ	Á	±	¡	‘		q	a	Q	A	1	!		ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡qaQA1!ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡‘qaQA1!°¡’ƒteVG8)üíÞÏ
	Line 482939: 	ñ	á	Ñ	Á	±	¡	‘		q	a	Q	A	1	!		ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡qaQA1!ñáÑÁ±¡‘qaQA1!ñáÑÁ±¡‘qaQA1!°¡’ƒteVG8)üíÞÏ
	Line 759315: ~LATLONGÄ~weak_ptr®
```

I know nothing about  .vs or sqlite.

**Comment by rgleason** [comment]
Pushed changes master branch  https://github.com/rgleason/AutoTrackRaymarine_pi
Bionic is failing on missing declaration "HANDLE"  
I think this comes from "src/serial"
I notice that OpenCPN has lib/serial   should I just use that?
What is the .vs directory for?
Douwe did not push a complete working plugin, it was missing the cmake directory and possibly something else.

**Comment by rgleason** [comment]
Windows build has  unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" 
and unresolved external symbol "public: __thiscall opencpn_plugin_117::opencpn_plugin_117(void *)"

Linux and Flatpak builds have unresolved "HANDLE"

**Comment by bdbcat** [comment]
Rick....
I think API117 is not working for you, somehow.
Please switch back to API 116, push to gihub, and show a link to the build logs.
Dave

**Comment by rgleason** [comment]
Ok, will do so, but the plugin requires 117

**Comment by rgleason** [comment]
What should I change, just the opencpn.lib and ocpn_plugin.h ?   Not the plugin src files.

**Comment by rgleason** [comment]
I can get it to build if I just use the original ocpn_plugin.h file  that had the first part and none of the extended waypoint modifications.

**Comment by bdbcat** [comment]
Rick...
1.  Do whatever it takes to build.  No need to change opencpn.lib.  You will need to change the ocpn_plugin.h, and the plugin src file:AutoTrackRaymarine_pi.cpp, line 75.
2. The "HANDLE" problem on Bionic":  This plugin, as coded, will only work on Windows.  Might as well adjust the CI files to build only with appveyor.  Nothing else will build, or be usable, without significant effort.

**Comment by rgleason** [comment]
I think I am going to stop on this for awhile. 
The old api117 ocpn_plugin.h is not building for windows now.
I don't understand why the new extended api117 with the correct opencpn.lib does not work. Shouldn't it?

**Comment by bdbcat** [comment]
It should, and it does for me in VS2017 Studio (not from command line).  So I do not know what is going wrong.
What is this plugin intended to do?  Who wrote it?

**Comment by rgleason** [comment]
I am using the command line. Will try VS2017.   This plugin is an autopilot for Raymarine EVO done by Douwe Fokkema (does extensive work on Radar_pi)

**Comment by bdbcat** [comment]
Sounds useful, even if "Windows only"...

**Comment by rgleason** [comment]
I must have changed something in my local environment so it does not build now.  I have tried several things using Visual Studio and command line, still fails with the errors above.
It built for windows earlier  https://ci.appveyor.com/project/rgleason/autotrackraymarine-pi/builds/41708809
Yesterday was the last commit that built https://github.com/rgleason/AutoTrackRaymarine_pi/commit/670076cea0dbaa0495967cdbf3967fcd1e2f4055

**Comment by rgleason** [comment]
Have gone back to the commit that built, and that fails too.  I think something has changed in the serial directory, so I am going to try libs/serial from OPenCPN.

**Comment by rgleason** [comment]
Moved src/serial to src/OLD/serial   Copied Opencpn/libs/serial to AutoTrackRaymarine/libs/serial
Then under  if(NOT OCPN_FLATPAK_CONFIG)
added     add_subdirectory(libs/serial)	  and      target_link_libraries(${PACKAGE_NAME} ocpn::serial)

Tried to build with comand line and VS.  Same errors which seem to come from this line in AutoTrackRaymarine_pi
```
Search "AutoTrackRaymarine_pi::AutoTrackRaymarine_pi" (1 hit in 1 file of 120 searched)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.cpp (1 hit)
	Line 74: AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *ppimgr)

OR  
AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *ppimgr)
  : opencpn_plugin_117(ppimgr)

```
a similar entry in climatology is

```
climatology_pi::climatology_pi(void *ppimgr)
      :opencpn_plugin_116(ppimgr)
```
So it looks ok.

**Comment by rgleason** [comment]
Also had updated opencpn.lib using the dropbox version from you. and updated ocpn_plugin.h to be the most current.

**Comment by rgleason** [comment]
There is is space after the ":"  maybe that is the problem?

**Comment by rgleason** [comment]
Not the problem.

```
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\libs\ocpn-api\ocpn_plugin.h (2 hits)
	Line 547: class DECL_EXP opencpn_plugin_117 : public opencpn_plugin_116 {
	Line 549:   opencpn_plugin_117(void *pmgr);

```
From ocpn_plugin.h file  Line 547

class DECL_EXP opencpn_plugin_117 : public opencpn_plugin_116 {
public:
  opencpn_plugin_117(void *pmgr);
  /*
   * Forms a semantic version together with GetPlugInVersionMajor() and
   * GetPlugInVersionMinor(), see https://semver.org/
   */
  virtual int GetPlugInVersionPatch();

  /** Post-release version part, extends the semver spec. */
  virtual int GetPlugInVersionPost();

  /** Pre-release tag version part, see GetPlugInVersionPatch() */
  virtual const char *GetPlugInVersionPre();

  /** Build version part  see GetPlugInVersionPatch(). */
  virtual const char *GetPlugInVersionBuild();

```
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\libs\ocpn-api\ocpn_plugin.h (3 hits)
	Line 554:   virtual int GetPlugInVersionPatch();
	Line 559:   /** Pre-release tag version part, see GetPlugInVersionPatch() */
	Line 562:   /** Build version part  see GetPlugInVersionPatch(). */
```

**Comment by rgleason** [comment]
Yet there seems to be some error with regard to  
unresolved external symbol
    opencpn_plugin_117::opencpn_plugin_117(void *)    and
    opencpn_plugin_117::GetPlugInVersionPatch(void)

```
     Creating library C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.lib and object C:/Users/fcgle/source/AutoTrackRaymarine_pi/build     Creating library C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.lib and object C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.exp
AutoTrackRaymarine_pi.obj : error LNK2019: unresolved external symbol "public: __thiscall opencpn_plugin_117::opencpn_plugin_117(void *)" (??0opencpn_plugin_117@@QAE@PAX@Z) referenced in function "public: __thiscall AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *)" (??0AutoTrackRaymarine_pi@@QAE@PAX@Z) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
NGT1Read.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
SerialPort.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
AutoTrackRaymarine_pi.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
PreferencesDialog.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)"/Release/AutoTrackRaymarine_pi.exp
AutoTrackRaymarine_pi.obj : error LNK2019: unresolved external symbol "public: __thiscall opencpn_plugin_117::opencpn_plugin_117(void *)" (??0opencpn_plugin_117@@QAE@PAX@Z) referenced in function "public: __thiscall AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *)" (??0AutoTrackRaymarine_pi@@QAE@PAX@Z) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
NGT1Read.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
SerialPort.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
AutoTrackRaymarine_pi.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)" (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
PreferencesDialog.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)"
```

**Comment by rgleason** [comment]
What is GetPluginVersionPatch used for?

**Comment by rgleason** [comment]
I just found in version.h.in under cmake/in-files/

```
#define PLUGIN_NAME ${PROJECT_NAME}
#define PLUGIN_CATALOG_NAME wxS("opencpn-${PACKAGE_NAME}")
#define PLUGIN_VERSION_MAJOR ${VERSION_MAJOR}
#define PLUGIN_VERSION_MINOR ${VERSION_MINOR}
#define PLUGIN_VERSION_PATCH ${VERSION_PATCH}    <----- SO IT EXISTS!!!!
#define PLUGIN_VERSION_TWEAK ${VERSION_TWEAK}
#define PLUGIN_VERSION_DATE "${VERSION_DATE}"
#define PLUGIN_COMMON_NAME "${COMMON_NAME}"
```

**Comment by rgleason** [comment]
Is this a problem with TP Templates?  Perhaps not being configured for API 117?

**Comment by rgleason** [comment]
Now using back2working branch
https://github.com/rgleason/AutoTrackRaymarine_pi/tree/back2working

**Comment by jongough** [comment]
Frontend2 (testplugin_pi) is OCPN API agnostic, i.e. the build process does not set or dictate an API version, rather this is up to the plugin developer/maintainer. To build you will need to provide the header and lib files needed for the particular API version and ensure the changes to the definition of the plugin in the plugin header file and main source file match. There is no change to any of the fontend2 files and only, possibly, a change in the plugin specific part of the CMakeLists.txt file.

**Comment by jongough** [comment]
There appear to be missing header files. Just tried the back2woking branch and both 'conio.h' and 'info.h' are not found in the include files, 'Info.h' is in the 'src' directory but the case of the first letter is different, i.e 'info.h' vs 'Info.h'. 'HANDLE' is a missing definition that is needed. So there is a need to go back to the point that the files and definitions are available and see what happened to them.

**Comment by jongough** [comment]
I have gone back to the source repository,  douwefokkema/AutoTrackRaymarine_pi, and it would appear that this is not a complete repository. The CMakeLists.txt file refers to 'cmake/*' files but the 'cmake' directory does not seem to exist in either branch (master or api10).

So it appears that there are other issues here that need sorting out first. The developer should be contacted to find out what has happened to their repository.

**Comment by rgleason** [comment]
It looks like NGT1Read.cpp is using a small "i" !!
```
Search "Info.h" (5 hits in 4 files of 120 searched)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.cpp (1 hit)
	Line 39: #include "Info.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\Info.cpp (1 hit)
	Line 29: #include "Info.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\NGT1Read.cpp (1 hit)
	Line 34: #include "info.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\OLD\serial\src\impl\list_ports\list_ports_osx.cc (2 hits)
	Line 251:         port_info.hardware_id = "n/a";
	Line 276:                 port_info.hardware_id = cstring;
```

**Comment by rgleason** [comment]
fixed the "i"  and added int GetPlugInVersionPatch();  to the AutoTrackRaymarine_pi.h file.
still does not build but getting closer. Pushed to the optimistic  [back2working branch ](https://github.com/rgleason/AutoTrackRaymarine_pi/tree/back2working)

**Comment by rgleason** [comment]
Still getting
```
     Creating library C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.lib and object C:/Users/fcgle/source/AutoTrackRaymarine_pi/build/Release/AutoTrackRaymarine_pi.exp
AutoTrackRaymarine_pi.obj : 
error LNK2019: unresolved external symbol "public: __thiscall opencpn_plugin_117::opencpn_plugin_117(void *)" (??0opencpn_plugin_117@@QAE@PAX@Z) 
referenced in function "public: __thiscall AutoTrackRaymarine_pi::AutoTrackRaymarine_pi(void *)" (??0AutoTrackRaymarine_pi@@QAE@PAX@Z) [C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]

NGT1Read.obj : error LNK2001: unresolved external symbol "public: virtual int __thiscall opencpn_plugin_117::GetPlugInVersionPatch(void)"
 (?GetPlugInVersionPatch@opencpn_plugin_117@@UAEHXZ) 
[C:\Users\fcgle\source\AutoTrackRaymarine_pi\build\AutoTrackRaymarine_pi.vcxproj]
```

**Comment by jongough** [comment]
Why are you using 1.17 of the OCPN API? Is there something that is needed in that version?
If you are building on windows it would appear that you are using a version 1.16 or earlier OCPN lib file hence the API is not defined correctly for OCPN and it will fail to link.

**Comment by rgleason** [comment]
The plugin uses and requires the first part of  api117 before the waypoint extension. I believe I am using the most current opencpn.lib.

**Comment by rgleason** [comment]
Ps. It is the opencpn.lib file that bdbcat left a dropbox link above.

**Comment by jongough** [comment]
The problem is in Appveyor it downloads opencpn.lib from [https://downloads.sourceforge.net/project/opencpnplugins/opencpn_lib/4.99.1405-vc141_xp/opencpn.lib](https://downloads.sourceforge.net/project/opencpnplugins/opencpn_lib/4.99.1405-vc141_xp/opencpn.lib) and places this in the build directory which gets used before the version you are pointing too. There should be no need for a compile of OCPN to get the lib file, it is supposed to be supplied as a matching pair with the ocpn_plugin.h, but...

I am not sure who is maintaining the sourceforge downloads, but they need to update the version there to make things work.

**Comment by rgleason** [comment]
Jon, thank you for thinking of this and the sourceforge file probably does need an update, but I've been building locally with what I think is the api117 file pair  ocpn-plugin.h  and opencpn.lib  from the OpenCPN version 5.6.0  and it is still failing for me.

Perhaps I'll be able to figure this out.  I was building earlier but something happened.

**Comment by bdbcat** [comment]
Rick....
Jon is on the right track, I think.
"Fortunately", the file referenced 
https://downloads.sourceforge.net/project/opencpnplugins/opencpn_lib/4.99.1405-vc141_xp/opencpn.lib
does not exist.  So the build seems to carry on and use the version bundled in the github tree.

To be very sure that we have the correct opencpn.lib file, I have uploaded a freshly built version tracking O56, with API 117.  Here it is:
https://www.dropbox.com/s/z1jk62p5m3yyti6/opencpn.lib?dl=0
Please get and try your local build with this file.
Dave

**Comment by jongough** [comment]
Dave,
   Is the downloads.sourceforge still being supported? Or have these files been moved somewhere else. I need to update the appveyor.yml for frontend2 if it has or will change.

**Comment by rgleason** [comment]
Using that file from Dave's download, causes it to build the executables and tarballs!
It is deployed to Cloudsmith now for Windows only. It is not yet finished but can be tested
https://github.com/rgleason/AutoTrackRaymarine_pi
Thank you Dave and Jon.  I will make a PR to Douwe and maybe improve it further.

**Comment by rgleason** [comment]
Dave,
I updated autotrackraymarine_pi to conform to Jon's very recent push last night, to testplugin.
There were two changes,  to ocpn-plugins.h  and opencpn.lib  under the libs directory.
Afterwards autotrackraymarine_pi would not compile and complete.

Therefore I went back to using the version of opencpn.lib that you[ attached above](https://github.com/OpenCPN/plugins/issues/511#issuecomment-987547006)

and the ocpn-plugin.h directly from the Opencpn Version 5.6.0 github repository

and that Works!

Which versions should Jon be using in Testplugin?
